design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+  Object getObject(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an Object.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Object getObject(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  String getString(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a string.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  String getString(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Integer getInt(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as an integer.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Integer getInt(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Long getLong(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Long getLong(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Double getDouble(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Double getDouble(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
+```
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
```

Comment:
I originally commented this above, but I see that you've thought this through. I tend to disagree here - having strings as column types seems somewhat annoying. Every time I want to programatically get a field I'll need to parse the type and that's not something that's easy (especially for Structs). It will also make it so that each location that is using types will need to re-parse it, instead of having the parsing done once when we construct the type and pass it around alongside the Row object. This will also make it harder for us to adjust our type system because we could accidentally break third-party parsers when maintaining backwards compatibility.

In general, exposing "String" for data types is always super error prone - I'd urge against it if possible. At a minimum, we should wrap the string in a `ColumnType` class so that later we can add methods that will do the parsing for the user.

Reply:
To be honest I suspect very few users will care about the types of the columns, at least they're not going to introspect them before looking up fields. When writing an app the app developer almost always knows the schema up front, i.e. they know that column 3 of their query result is a boolean. It's quite unusual for an app to have to work against a completely unknown schema.
Having said I do agree that string aren't great for column types (probably better to use an enum here), just that most people won't look at them anyway in normal use.

Reply:
I second what @agavra says.

Reply:
I also agree with @agavra. I feel that one of the benefits of shipping our own client is that we can expose this kind of information in a strongly typed way, which is more conducive to helping users write robust, correct code.

Reply:
Just to clarify the point I was trying to make:

I agree that we shouldn't use strings for column types - an enum is better there.

But.... @agavra said:

> Every time I want to programatically get a field I'll need to parse the type and that's not something that's easy (especially for Structs).

This is not true. In vast majority of cases the user won't ever look at the method that returns the type of the columns, as they will already know the schema. So whether they return strings or not is probably only of interest to a very small set of developers who are writing an app against a completely unknown schema.

Does that make sense?

Reply:
> In vast majority of cases the user won't ever look at the method that returns the type of the columns, as they will already know the schema. 

Then why have this method at all? It seems that the people who want it won't be able to use it (the ones who want to programmatically access it) and the others won't need it. Without structured types, it seems useful only for debugging/logging.

Reply:
I think the metadata is only useful for some niche reflection/introspection type use cases. An analogy is calling a Java method you don't need to use the reflection API to call a method most of the time, because you know the method signature when you write the code.

I'd recommend leaving the method there as it might benefit some cases, and it's easy to implement using enums. Structured types for structs, arrays, maps etc have little value imho, even for the reflection use cases. Saying a struct is a Json object or a field is just a decimal is good enough. Note that JDBC doesn't feel the need to make types more structured than simple ones, nor does Calcite and they do fine.

Reply:
Imho an enum is ok, and a new Interface is a bit overkill. But not a showstopper.

Reply:
Because changing return types of methods is not easily possible (ie, cannot deprecate the old method, as add a new one with the same name but only different return type), I would opt for the `ColumnType` interface as it's more future proof.